### PR TITLE
[backend] Improve performance of taxii/stream data loading by batching element complete resolution

### LINKED ISSUE #3250: Improve performance of taxii/stream data loading by batching element refs resolution
## Use case

Improve performance of taxii/stream data loading by batching element refs resolution
Currently stix loading is done in a loop => we can batch load and then redispatch the results

### PR DESCRIPTION
**Pull request by richard-julien**

### CODE REVIEW

File: opencti-platform/opencti-graphql/src/database/middleware.js [hunk 1]
```diff
@@ -392,90 +392,124 @@ export const loadEntity = async (context, user, entityTypes, args = {}) => {
...
-  const refArgs = { indices: [READ_INDEX_STIX_META_RELATIONSHIPS, READ_INDEX_STIX_CYBER_OBSERVABLE_RELATIONSHIPS], fromId: elementId };
+  const refArgs = { indices: [READ_INDEX_STIX_META_RELATIONSHIPS, READ_INDEX_STIX_CYBER_OBSERVABLE_RELATIONSHIPS], fromId: elementIds };
   const refsRelations = await listAllRelations(context, user, relTypes, refArgs);
-  const data = {};
+  const refsPerElements = R.groupBy((r) => r.fromId, refsRelations);
   // Parallel resolutions
   const toResolvedIds = R.uniq(refsRelations.map((rel) => rel.toId));
-  const toResolved = await elFindByIds(context, user, toResolvedIds, { toMap: true });
-  if (refsRelations.length > 0) {
+  // const toResolved = await elFindByIds(context, user, toResolvedIds, { toMap: true });
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/database/middleware.js]: This line can be removed ?

File: opencti-platform/opencti-graphql/src/database/middleware.js [hunk 2]
```diff
@@ -392,90 +392,124 @@ export const loadEntity = async (context, user, entityTypes, args = {}) => {
...
+  const toResolvedElements = await elFindByIds(context, user, toResolvedIds, { withoutRels: true, connectionFormat: false });
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/database/middleware.js]: Can we used toMap: true on args ?

File: opencti-platform/opencti-graphql/src/database/middleware.js [hunk 3]
```diff
@@ -392,90 +392,124 @@ export const loadEntity = async (context, user, entityTypes, args = {}) => {
...
-  return undefined;
+  return instance ? convertStoreToStix(instance) : undefined;
 };
 export const stixLoadByIds = async (context, user, ids, opts = {}) => {
-  const instances = await Promise.map(ids, (id) => stixLoadById(context, user, id, opts), {
-    concurrency: ES_MAX_CONCURRENCY,
-  });
-  return instances.filter((i) => isNotEmptyField(i));
+  const elements = await loadByIdsWithDependencies(context, user, ids, { ...opts, onlyMarking: false, withoutRels: true });
+  return elements.map((instance) => convertStoreToStix(instance));
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/database/middleware.js:510]: Why for a single ID we want the rel, and for multiple IDs we won't ?
- Single ID -> storeLoadByIdWithRefs
- Multiple ID -> withoutRels: true

File: opencti-platform/opencti-graphql/src/graphql/sseMiddleware.js [hunk 4]
```diff
@@ -371,21 +374,20 @@ const createSseMiddleware = () => {
     if (noDependencies === false) {
       const allRelCallback = async (relations) => {
         const notCachedRelations = relations.filter((m) => !cache.has(m.standard_id));
-        for (let relIndex = 0; relIndex < notCachedRelations.length; relIndex += 1) {
-          const relation = notCachedRelations[relIndex];
+        const findRelIds = notCachedRelations.map((r) => r.internal_id);
+        const missingRelations = await storeLoadByIdsWithRefs(context, req.user, findRelIds);
+        for (let relIndex = 0; relIndex < missingRelations.length; relIndex += 1) {
+          const missingRelation = missingRelations[relIndex];
           if (channel.connected()) {
```
Comment (RomuDeuxfois) [inline opencti-platform/opencti-graphql/src/graphql/sseMiddleware.js:381]: && !cache.has(stixRelation.id) ?

File: opencti-platform/opencti-graphql/src/database/middleware.js (hunk 1 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/database/middleware.js]: yes

File: opencti-platform/opencti-graphql/src/database/middleware.js (hunk 2 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/database/middleware.js]: Yes should be possible

File: opencti-platform/opencti-graphql/src/database/middleware.js (hunk 3 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/database/middleware.js:510]: storeLoadByIdWithRefs call storeLoadByIdsWithRefs under the hood
But stixLoadByIds should use storeLoadByIdsWithRefs

File: opencti-platform/opencti-graphql/src/graphql/sseMiddleware.js (hunk 4 above)
Comment (richard-julien) [reply opencti-platform/opencti-graphql/src/graphql/sseMiddleware.js:381]: all rels fetched already filtered from the cache.

> const notCachedRelations = relations.filter((m) => !cache.has(m.standard_id));

**Review by RomuDeuxfois — APPROVED** [review]

**Pull request merged**